File: sdk/flutter/ios/OpenStorySDK/Internal/Cache/ViewedStoryStateRepository.swift
import Foundation

internal final class ViewedStoryStateRepository: @unchecked Sendable {
    private let database: OpenStoryDatabase
    private let lock = NSLock()
    private var isLoaded = false
    private var cachedSnapshot = ViewedStoryStateSnapshot.empty

    init(database: OpenStoryDatabase) {
        self.database = database
    }

    func snapshot() -> ViewedStoryStateSnapshot {
        ensureLoaded()
        return currentSnapshot()
    }

    func currentSnapshot() -> ViewedStoryStateSnapshot {
        lock.lock()
        defer { lock.unlock() }
        return cachedSnapshot
    }

    func markViewed(_ storyRevisionId: String) -> Bool {
        let trimmed = storyRevisionId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return false
        }

        ensureLoaded()

        let inserted = lock.withLock { () -> Bool in
            if cachedSnapshot.viewedStoryRevisionIds.contains(trimmed) {
                return false
            }
            cachedSnapshot = cachedSnapshot.withViewedStory(trimmed)
            return true
        }

        guard inserted else {
            return false
        }

        do {
            try database.upsertViewedStoryRevision(
                storyRevisionId: trimmed,
                firstViewedAtEpochMs: Int64(Date().timeIntervalSince1970 * 1_000)
            )
        } catch {
            return false
        }

        return true
    }

    private func ensureLoaded() {
        let shouldLoad = lock.withLock { !isLoaded }

        guard shouldLoad else {
            return
        }

        let viewedStoryRevisionIds = Set(database.allViewedStoryRevisionIds())

        lock.withLock {
            if !isLoaded {
                cachedSnapshot = ViewedStoryStateSnapshot(
                    viewedStoryRevisionIds: viewedStoryRevisionIds
                )
                isLoaded = true
            }
        }
    }
}

private extension NSLock {
    func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock()
        defer { unlock() }
        return try body()
    }
}

File: sdk/flutter/android/src/main/kotlin/com/openstory/sdk/internal/cache/ViewedStoryStateRepository.kt
package com.openstory.sdk.internal.cache

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

internal class ViewedStoryStateRepository(
    private val dao: ViewedStoryRevisionDao,
) {
    private val mutex = Mutex()

    @Volatile
    private var isLoaded = false

    @Volatile
    private var cachedSnapshot = ViewedStoryStateSnapshot.EMPTY

    suspend fun snapshot(): ViewedStoryStateSnapshot {
        ensureLoaded()
        return cachedSnapshot
    }

    fun currentSnapshot(): ViewedStoryStateSnapshot {
        return cachedSnapshot
    }

    suspend fun markViewed(storyRevisionId: String): Boolean {
        if (storyRevisionId.isBlank()) {
            return false
        }

        ensureLoaded()

        val inserted = mutex.withLock {
            if (cachedSnapshot.viewedStoryRevisionIds.contains(storyRevisionId)) {
                false
            } else {
                cachedSnapshot = cachedSnapshot.withViewedStory(storyRevisionId)
                true
            }
        }

        if (inserted) {
            dao.upsert(
                ViewedStoryRevisionEntity(
                    storyRevisionId = storyRevisionId,
                    firstViewedAtEpochMs = System.currentTimeMillis(),
                ),
            )
        }

        return inserted
    }

    private suspend fun ensureLoaded() {
        if (isLoaded) {
            return
        }

        mutex.withLock {
            if (isLoaded) {
                return
            }

            cachedSnapshot = ViewedStoryStateSnapshot(
                viewedStoryRevisionIds = dao.allViewedStoryRevisionIds().toSet(),
            )
            isLoaded = true
        }
    }
}
